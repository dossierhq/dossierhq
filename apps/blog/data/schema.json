{
  "entityTypes": [
    {
      "name": "Article",
      "adminOnly": false,
      "authKeyPattern": "none",
      "fields": [
        {
          "name": "title",
          "type": "String",
          "list": false,
          "required": true,
          "adminOnly": false,
          "isName": true,
          "multiline": false,
          "matchPattern": null,
          "index": null
        },
        {
          "name": "slug",
          "type": "String",
          "list": false,
          "required": true,
          "adminOnly": false,
          "isName": false,
          "multiline": false,
          "matchPattern": "slug",
          "index": "articleSlug"
        },
        {
          "name": "body",
          "type": "RichText",
          "list": false,
          "required": true,
          "adminOnly": false,
          "richTextNodes": [
            "root",
            "paragraph",
            "text",
            "linebreak",
            "list",
            "listitem",
            "code",
            "code-highlight",
            "entityLink",
            "heading",
            "valueItem"
          ],
          "entityTypes": [],
          "linkEntityTypes": [
            "GlossaryTerm",
            "Article"
          ],
          "valueTypes": [
            "CloudinaryImage"
          ]
        }
      ]
    },
    {
      "name": "Chapter",
      "adminOnly": false,
      "authKeyPattern": "none",
      "fields": [
        {
          "name": "items",
          "type": "ValueItem",
          "list": true,
          "required": true,
          "adminOnly": false,
          "valueTypes": [
            "ArticleTocItem",
            "TocItem"
          ]
        }
      ]
    },
    {
      "name": "GlossaryTerm",
      "adminOnly": false,
      "authKeyPattern": "none",
      "fields": [
        {
          "name": "title",
          "type": "String",
          "list": false,
          "required": true,
          "adminOnly": false,
          "isName": true,
          "multiline": false,
          "matchPattern": null,
          "index": null
        },
        {
          "name": "slug",
          "type": "String",
          "list": false,
          "required": true,
          "adminOnly": false,
          "isName": false,
          "multiline": false,
          "matchPattern": "slug",
          "index": "glossarySlug"
        },
        {
          "name": "description",
          "type": "RichText",
          "list": false,
          "required": true,
          "adminOnly": false,
          "richTextNodes": [
            "root",
            "paragraph",
            "text",
            "linebreak",
            "entityLink"
          ],
          "entityTypes": [],
          "linkEntityTypes": [
            "GlossaryTerm"
          ],
          "valueTypes": []
        }
      ]
    }
  ],
  "valueTypes": [
    {
      "name": "ArticleTocItem",
      "adminOnly": false,
      "fields": [
        {
          "name": "title",
          "type": "String",
          "list": false,
          "required": true,
          "adminOnly": false,
          "isName": false,
          "multiline": false,
          "matchPattern": null,
          "index": null
        },
        {
          "name": "article",
          "type": "Entity",
          "list": false,
          "required": true,
          "adminOnly": false,
          "entityTypes": [
            "Article"
          ]
        }
      ]
    },
    {
      "name": "CloudinaryImage",
      "adminOnly": false,
      "fields": [
        {
          "name": "publicId",
          "type": "String",
          "list": false,
          "required": true,
          "adminOnly": false,
          "isName": false,
          "multiline": false,
          "matchPattern": null,
          "index": null
        },
        {
          "name": "width",
          "type": "Number",
          "list": false,
          "required": true,
          "adminOnly": false,
          "integer": true
        },
        {
          "name": "height",
          "type": "Number",
          "list": false,
          "required": true,
          "adminOnly": false,
          "integer": true
        },
        {
          "name": "alt",
          "type": "String",
          "list": false,
          "required": false,
          "adminOnly": false,
          "isName": false,
          "multiline": false,
          "matchPattern": null,
          "index": null
        }
      ]
    },
    {
      "name": "TocItem",
      "adminOnly": false,
      "fields": [
        {
          "name": "title",
          "type": "String",
          "list": false,
          "required": true,
          "adminOnly": false,
          "isName": false,
          "multiline": false,
          "matchPattern": null,
          "index": null
        },
        {
          "name": "items",
          "type": "ValueItem",
          "list": true,
          "required": true,
          "adminOnly": false,
          "valueTypes": [
            "ArticleTocItem",
            "TocItem"
          ]
        }
      ]
    }
  ],
  "patterns": [
    {
      "name": "none",
      "pattern": "^none$"
    },
    {
      "name": "slug",
      "pattern": "^[a-z0-9-]+$"
    }
  ],
  "indexes": [
    {
      "name": "articleSlug",
      "type": "unique"
    },
    {
      "name": "glossarySlug",
      "type": "unique"
    }
  ]
}
