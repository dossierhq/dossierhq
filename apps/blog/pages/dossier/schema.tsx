import dynamic from 'next/dynamic';
import type { JSX } from 'react';

const SchemaEditorPage = dynamic(
  () => import('../../components/SchemaEditorPage/SchemaEditorPage'),
  { ssr: false },
);

export default function SchemaEditorPage_(): JSX.Element {
  return <SchemaEditorPage />;
}
