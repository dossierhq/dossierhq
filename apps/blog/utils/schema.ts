import type { AdminSchemaSpecificationUpdate } from '@jonasb/datadata-core';
import { FieldType, RichTextNodeType } from '@jonasb/datadata-core';

export const schemaSpecification: AdminSchemaSpecificationUpdate = {
  entityTypes: [
    {
      name: 'GlossaryTerm',
      authKeyPattern: 'none',
      fields: [
        {
          name: 'title',
          type: FieldType.String,
          isName: true,
          required: true,
        },
        {
          name: 'slug',
          type: FieldType.String,
          index: 'glossarySlug',
          matchPattern: 'slug',
          required: true,
        },
        {
          name: 'description',
          type: FieldType.RichText,
          required: true,
          richTextNodes: [
            RichTextNodeType.root,
            RichTextNodeType.paragraph,
            RichTextNodeType.text,
            RichTextNodeType.entityLink,
          ],
          linkEntityTypes: ['GlossaryTerm'],
        },
      ],
    },
  ],
  indexes: [{ name: 'glossarySlug', type: 'unique' }],
  patterns: [
    { name: 'none', pattern: '^none$' },
    { name: 'slug', pattern: '^[a-z0-9-]+$' },
  ],
};
