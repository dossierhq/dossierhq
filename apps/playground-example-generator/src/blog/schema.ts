import type { AdminSchemaSpecificationUpdate } from '@jonasb/datadata-core';
import { RichTextNodeType } from '@jonasb/datadata-core';
import { FieldType } from '@jonasb/datadata-core';

export const SCHEMA: AdminSchemaSpecificationUpdate = {
  entityTypes: [
    {
      name: 'BlogPost',
      authKeyPattern: 'none',
      fields: [
        { name: 'title', type: FieldType.String, required: true, isName: true },
        {
          name: 'slug',
          type: FieldType.String,
          required: true,
          index: 'slug',
          matchPattern: 'slug',
        },
        {
          name: 'heroImage',
          type: FieldType.ValueItem,
          required: true,
          valueTypes: ['CloudinaryImage'],
        },
        {
          name: 'description',
          type: FieldType.RichText,
          required: true,
          richTextNodes: [
            RichTextNodeType.root,
            RichTextNodeType.paragraph,
            RichTextNodeType.text,
            RichTextNodeType.linebreak,
          ],
        },
        {
          name: 'body',
          type: FieldType.RichText,
          required: true,
          richTextNodes: [
            RichTextNodeType.root,
            RichTextNodeType.paragraph,
            RichTextNodeType.text,
            RichTextNodeType.linebreak,
            RichTextNodeType.entityLink,
            RichTextNodeType.valueItem,
          ],
          linkEntityTypes: ['BlogPost'],
          valueTypes: ['CloudinaryImage'],
        },
        {
          name: 'authors',
          type: FieldType.Entity,
          list: true,
          required: true,
          entityTypes: ['Person'],
        },
        { name: 'tags', type: FieldType.String, list: true },
      ],
    },
    {
      name: 'Person',
      authKeyPattern: 'none',
      fields: [{ name: 'title', type: FieldType.String, required: true, isName: true }],
    },
  ],
  valueTypes: [
    {
      name: 'CloudinaryImage',
      fields: [
        { name: 'publicId', type: FieldType.String, required: true },
        { name: 'width', type: FieldType.Number, integer: true, required: true },
        { name: 'height', type: FieldType.Number, integer: true, required: true },
        { name: 'alt', type: FieldType.String },
      ],
    },
  ],
  patterns: [
    {
      name: 'none',
      pattern: '^none$',
    },
    {
      name: 'slug',
      pattern: '^[a-z0-9-]+$',
    },
  ],
  indexes: [
    {
      name: 'slug',
      type: 'unique',
    },
  ],
};
