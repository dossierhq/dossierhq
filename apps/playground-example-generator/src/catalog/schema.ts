import type { AdminSchemaSpecificationUpdate } from '@dossierhq/core';
import { FieldType, RichTextNodeType } from '@dossierhq/core';

export const SCHEMA: AdminSchemaSpecificationUpdate = {
  entityTypes: [
    {
      name: 'BooleansEntity',
      fields: [
        { name: 'normal', type: FieldType.Boolean },
        { name: 'required', type: FieldType.Boolean, required: true },
      ],
    },
    {
      name: 'EntitiesEntity',
      fields: [
        { name: 'normal', type: FieldType.Entity },
        { name: 'required', type: FieldType.Entity, required: true },
        { name: 'list', type: FieldType.Entity, list: true },
        { name: 'stringsEntity', type: FieldType.Entity, entityTypes: ['StringsEntity'] },
        {
          name: 'stringsEntityList',
          type: FieldType.Entity,
          list: true,
          entityTypes: ['StringsEntity'],
        },
        {
          name: 'stringsAndLocationsEntity',
          type: FieldType.Entity,
          entityTypes: ['LocationsEntity', 'StringsEntity'],
        },
        {
          name: 'stringsAndLocationsEntityList',
          type: FieldType.Entity,
          list: true,
          entityTypes: ['LocationsEntity', 'StringsEntity'],
        },
      ],
    },
    {
      name: 'LocationsEntity',
      fields: [
        { name: 'normal', type: FieldType.Location },
        { name: 'required', type: FieldType.Location, required: true },
        { name: 'list', type: FieldType.Location, list: true },
        { name: 'requiredList', type: FieldType.Location, list: true, required: true },
      ],
    },
    {
      name: 'NumbersEntity',
      fields: [
        { name: 'normal', type: FieldType.Number },
        { name: 'required', type: FieldType.Number, required: true },
        { name: 'integer', type: FieldType.Number, integer: true },
        { name: 'list', type: FieldType.Number, list: true },
        { name: 'requiredList', type: FieldType.Number, list: true, required: true },
        {
          name: 'requiredIntegerList',
          type: FieldType.Number,
          list: true,
          required: true,
          integer: true,
        },
      ],
    },
    {
      name: 'RichTextsEntity',
      fields: [
        { name: 'normal', type: FieldType.RichText },
        { name: 'required', type: FieldType.RichText, required: true },
        {
          name: 'minimal',
          type: FieldType.RichText,
          richTextNodes: [
            RichTextNodeType.root,
            RichTextNodeType.paragraph,
            RichTextNodeType.text,
            RichTextNodeType.linebreak,
          ],
        },
        { name: 'list', type: FieldType.RichText, list: true },
        { name: 'adminOnly', type: FieldType.RichText, adminOnly: true },
        {
          name: 'stringsEntity',
          type: FieldType.RichText,
          richTextNodes: [
            RichTextNodeType.root,
            RichTextNodeType.paragraph,
            RichTextNodeType.text,
            RichTextNodeType.linebreak,
            RichTextNodeType.entity,
          ],
          entityTypes: ['StringsEntity'],
        },
        {
          name: 'numbersEntityLink',
          type: FieldType.RichText,
          richTextNodes: [
            RichTextNodeType.root,
            RichTextNodeType.paragraph,
            RichTextNodeType.text,
            RichTextNodeType.linebreak,
            RichTextNodeType.entityLink,
          ],
          linkEntityTypes: ['NumbersEntity'],
        },
        {
          name: 'nestedValueItem',
          type: FieldType.RichText,
          richTextNodes: [
            RichTextNodeType.root,
            RichTextNodeType.paragraph,
            RichTextNodeType.text,
            RichTextNodeType.linebreak,
            RichTextNodeType.valueItem,
          ],
          valueTypes: ['NestedValueItem'],
        },
      ],
    },
    {
      name: 'StringsEntity',
      nameField: 'title',
      fields: [
        { name: 'title', type: FieldType.String },
        { name: 'normal', type: FieldType.String },
        { name: 'required', type: FieldType.String, required: true },
        { name: 'matchPattern', type: FieldType.String, matchPattern: 'fooBarBaz' },
        { name: 'list', type: FieldType.String, list: true },
        { name: 'requiredList', type: FieldType.String, list: true, required: true },
        {
          name: 'requiredListMatchPattern',
          type: FieldType.String,
          list: true,
          required: true,
          matchPattern: 'fooBarBaz',
        },
      ],
    },
    {
      name: 'ValueItemsEntity',
      fields: [
        { name: 'normal', type: FieldType.ValueItem },
        { name: 'required', type: FieldType.ValueItem, required: true },
        { name: 'list', type: FieldType.ValueItem, list: true },
        { name: 'requiredList', type: FieldType.ValueItem, list: true, required: true },
        { name: 'adminOnly', type: FieldType.ValueItem, adminOnly: true },
        { name: 'cloudinaryImage', type: FieldType.ValueItem, valueTypes: ['CloudinaryImage'] },
      ],
    },
  ],
  valueTypes: [
    {
      name: 'CloudinaryImage',
      fields: [
        { name: 'publicId', type: FieldType.String, required: true },
        { name: 'width', type: FieldType.Number, integer: true, required: true },
        { name: 'height', type: FieldType.Number, integer: true, required: true },
        { name: 'alt', type: FieldType.String },
      ],
    },
    {
      name: 'NestedValueItem',
      fields: [
        { name: 'text', type: FieldType.String },
        { name: 'child', type: FieldType.ValueItem, valueTypes: ['NestedValueItem'] },
      ],
    },
    {
      name: 'StringsValueItem',
      fields: [
        { name: 'normal', type: FieldType.String },
        { name: 'required', type: FieldType.String, required: true },
        { name: 'matchPattern', type: FieldType.String, matchPattern: 'fooBarBaz' },
        { name: 'list', type: FieldType.String, list: true },
        { name: 'requiredList', type: FieldType.String, list: true, required: true },
        {
          name: 'requiredListMatchPattern',
          type: FieldType.String,
          list: true,
          required: true,
          matchPattern: 'fooBarBaz',
        },
      ],
    },
  ],
  patterns: [
    { name: 'none', pattern: '^none$' },
    { name: 'slug', pattern: '^[a-z0-9-]+$' },
    { name: 'fooBarBaz', pattern: '^(foo|bar|baz)$' },
  ],
  indexes: [{ name: 'slug', type: 'unique' }],
};
