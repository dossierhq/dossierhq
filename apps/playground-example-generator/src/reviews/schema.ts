import type { SchemaSpecificationUpdate } from '@dossierhq/core';
import { FieldType, REQUIRED_RICH_TEXT_NODES, RichTextNodeType } from '@dossierhq/core';

export const SCHEMA: SchemaSpecificationUpdate = {
  entityTypes: [
    {
      name: 'PersonalNote',
      authKeyPattern: 'subject',
      fields: [
        {
          name: 'note',
          type: FieldType.RichText,
          richTextNodes: [...REQUIRED_RICH_TEXT_NODES, RichTextNodeType.entity],
          entityTypes: ['PlaceOfBusiness'],
        },
        {
          name: 'placeOfBusiness',
          type: FieldType.Reference,
          entityTypes: ['PlaceOfBusiness'],
          required: true,
        },
      ],
    },
    {
      name: 'PlaceOfBusiness',
      nameField: 'name',
      fields: [
        { name: 'name', type: FieldType.String, required: true },
        { name: 'address', type: FieldType.Component, componentTypes: ['Address'] },
        { name: 'slogan', type: FieldType.String, required: true },
        { name: 'description', type: FieldType.String, multiline: true, required: true },
      ],
    },
    {
      name: 'Review',
      fields: [
        { name: 'reviewer', type: FieldType.Reference, entityTypes: ['Reviewer'], required: true },
        {
          name: 'placeOfBusiness',
          type: FieldType.Reference,
          entityTypes: ['PlaceOfBusiness'],
          required: true,
        },
        { name: 'review', type: FieldType.String, required: true },
      ],
    },
    {
      name: 'Reviewer',
      nameField: 'name',
      fields: [{ name: 'name', type: FieldType.String, required: true }],
    },
  ],
  componentTypes: [
    {
      name: 'Address',
      fields: [
        { name: 'location', type: FieldType.Location, required: true },
        { name: 'line1', type: FieldType.String, required: true },
        { name: 'line2', type: FieldType.String },
        { name: 'zip', type: FieldType.String, required: true },
        { name: 'city', type: FieldType.String, required: true },
      ],
    },
  ],
  patterns: [
    {
      name: 'subject',
      pattern: '^subject$',
    },
  ],
};
