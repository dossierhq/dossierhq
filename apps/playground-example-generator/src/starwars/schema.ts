import type { AdminSchemaSpecificationUpdate } from '@dossierhq/core';
import { FieldType } from '@dossierhq/core';

export const SCHEMA: AdminSchemaSpecificationUpdate = {
  entityTypes: [
    {
      name: 'Film',
      authKeyPattern: 'default',
      nameField: 'title',
      fields: [
        { name: 'title', type: FieldType.String, required: true },
        { name: 'episodeId', type: FieldType.Number, integer: true, required: true },
        { name: 'director', type: FieldType.String, required: true },
        { name: 'producers', type: FieldType.String, list: true, required: true },
        { name: 'releaseDate', type: FieldType.String, required: true },
        { name: 'openingCrawl', type: FieldType.String, required: true, multiline: true },
        {
          name: 'characters',
          type: FieldType.Entity,
          list: true,
          entityTypes: ['Person'],
          required: true,
        },
        {
          name: 'starships',
          type: FieldType.Entity,
          list: true,
          entityTypes: ['Starship'],
          required: true,
        },
        {
          name: 'vehicles',
          type: FieldType.Entity,
          list: true,
          entityTypes: ['Vehicle'],
          required: true,
        },
        {
          name: 'planets',
          type: FieldType.Entity,
          list: true,
          entityTypes: ['Planet'],
          required: true,
        },
        {
          name: 'species',
          type: FieldType.Entity,
          list: true,
          entityTypes: ['Species'],
          required: true,
        },
      ],
    },
    {
      name: 'Person',
      authKeyPattern: 'default',
      nameField: 'name',
      fields: [
        { name: 'name', type: FieldType.String, required: true },
        { name: 'gender', type: FieldType.String, required: true },
        { name: 'skinColors', type: FieldType.String, list: true, required: true },
        { name: 'hairColors', type: FieldType.String, list: true, required: true },
        { name: 'eyeColors', type: FieldType.String, list: true, required: true },
        { name: 'height', type: FieldType.String, required: true },
        { name: 'mass', type: FieldType.String, required: true },
        { name: 'homeworld', type: FieldType.Entity, entityTypes: ['Planet'], required: true },
        { name: 'birthYear', type: FieldType.String, required: true },
      ],
    },
    {
      name: 'Planet',
      authKeyPattern: 'default',
      nameField: 'name',
      fields: [
        { name: 'name', type: FieldType.String, required: true },
        { name: 'climate', type: FieldType.String, list: true, required: true },
        { name: 'surfaceWater', type: FieldType.String, required: true },
        { name: 'diameter', type: FieldType.String, required: true },
        { name: 'terrain', type: FieldType.String, list: true, required: true },
        { name: 'gravity', type: FieldType.String, required: true },
        { name: 'rotationPeriod', type: FieldType.String, required: true },
        { name: 'orbitalPeriod', type: FieldType.String, required: true },
        { name: 'population', type: FieldType.String, required: true },
      ],
    },
    {
      name: 'Species',
      authKeyPattern: 'default',
      nameField: 'name',
      fields: [
        { name: 'name', type: FieldType.String, required: true },
        { name: 'classification', type: FieldType.String, required: true },
        { name: 'designation', type: FieldType.String, required: true },
        { name: 'skinColors', type: FieldType.String, list: true, required: true },
        { name: 'hairColors', type: FieldType.String, list: true, required: true },
        { name: 'eyeColors', type: FieldType.String, list: true, required: true },
        { name: 'language', type: FieldType.String, required: true },
        { name: 'averageLifespan', type: FieldType.String, required: true },
        { name: 'averageHeight', type: FieldType.String, required: true },
        { name: 'people', type: FieldType.Entity, list: true, entityTypes: ['Person'] },
        { name: 'homeworld', type: FieldType.Entity, entityTypes: ['Planet'] },
      ],
    },
    {
      name: 'Starship',
      authKeyPattern: 'default',
      nameField: 'starshipClass',
      fields: [
        { name: 'starshipClass', type: FieldType.String, required: true },
        { name: 'mglt', type: FieldType.String, required: true },
        { name: 'hyperdriveRating', type: FieldType.String, required: true },
        { name: 'pilots', type: FieldType.Entity, list: true, entityTypes: ['Person'] },
      ],
    },
    {
      name: 'Transport',
      authKeyPattern: 'default',
      nameField: 'name',
      fields: [
        { name: 'name', type: FieldType.String, required: true },
        { name: 'model', type: FieldType.String, required: true },
        { name: 'manufacturers', type: FieldType.String, list: true, required: true },
        { name: 'consumables', type: FieldType.String, required: true },
        { name: 'cargoCapacity', type: FieldType.String, required: true },
        { name: 'crew', type: FieldType.String, required: true },
        { name: 'passengers', type: FieldType.String, required: true },
        { name: 'maxAtmospheringSpeed', type: FieldType.String, required: true },
        { name: 'length', type: FieldType.String, required: true },
        { name: 'costInCredits', type: FieldType.String, required: true },
      ],
    },
    {
      name: 'Vehicle',
      authKeyPattern: 'default',
      nameField: 'vehicleClass',
      fields: [
        { name: 'vehicleClass', type: FieldType.String, required: true },
        { name: 'pilots', type: FieldType.Entity, list: true, entityTypes: ['Person'] },
      ],
    },
  ],
  patterns: [{ name: 'default', pattern: '^$' }],
};
