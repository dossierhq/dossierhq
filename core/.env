DATABASE_URL=postgres://coretestuser:coretestpass@localhost:5432/datadata-core
