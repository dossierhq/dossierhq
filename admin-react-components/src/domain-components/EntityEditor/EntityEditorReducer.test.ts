import type { AdminEntity } from '@datadata/core';
import { CoreTestUtils } from '@datadata/core';
import schema from '../../stories/StoryboardSchema';
import { TestContextAdapter } from '../../test/TestContextAdapter';
import { foo1Id } from '../../test/EntityFixtures';
import type { EntityEditorState } from './EntityEditorReducer';
import {
  forTest,
  reduceEditorState,
  SetNameAction,
  UpdateEntityAction,
} from './EntityEditorReducer';
const { expectOkResult } = CoreTestUtils;
const { createEditorEntityState } = forTest;

function entityState(entity: AdminEntity): EntityEditorState['entity'] {
  const entitySpec = schema.getEntityTypeSpecification(entity._type);
  if (!entitySpec) {
    throw new Error('No entity spec');
  }
  return createEditorEntityState(entitySpec, entity);
}

function newState(state: { id: string } & Partial<EntityEditorState>): EntityEditorState {
  return { initMessage: null, entityLoadMessage: null, schema, entity: null, ...state };
}

function stateWithoutSchema(state: EntityEditorState) {
  const { schema, ...newState } = state;
  if (newState.entity) {
    const { entitySpec, ...newEntity } = newState.entity;
    newEntity.fields = newEntity.fields.map((field) => {
      const { fieldSpec, ...newField } = field;
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      return newField as any;
    });
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    newState.entity = newEntity as any;
  }
  return newState;
}

describe('reduceEditorState', () => {
  test('SetNameAction', async () => {
    const fixtures = new TestContextAdapter();
    const entityResult = await fixtures.getEntity(foo1Id);
    if (expectOkResult(entityResult)) {
      const state = reduceEditorState(
        newState({ id: foo1Id, entity: entityState(entityResult.value) }),
        new SetNameAction('New name')
      );
      expect(state.entity?.name).toEqual('New name');
    }
  });

  test('UpdateEntityAction', async () => {
    const fixtures = new TestContextAdapter();
    const entityResult = await fixtures.getEntity(foo1Id);
    if (expectOkResult(entityResult)) {
      const entity = entityResult.value;
      const state = reduceEditorState(newState({ id: entity.id }), new UpdateEntityAction(entity));
      expect(stateWithoutSchema(state)).toMatchInlineSnapshot(`
        Object {
          "entity": Object {
            "fields": Array [
              Object {
                "initialValue": "Hello",
                "value": "Hello",
              },
              Object {
                "initialValue": Array [
                  "one",
                  "two",
                  "three",
                ],
                "value": Array [
                  "one",
                  "two",
                  "three",
                ],
              },
              Object {
                "initialValue": Object {
                  "lat": 55.60498,
                  "lng": 13.003822,
                },
                "value": Object {
                  "lat": 55.60498,
                  "lng": 13.003822,
                },
              },
              Object {
                "initialValue": Array [
                  Object {
                    "lat": 55.60498,
                    "lng": 13.003822,
                  },
                  Object {
                    "lat": 56.381561,
                    "lng": 13.99286,
                  },
                ],
                "value": Array [
                  Object {
                    "lat": 55.60498,
                    "lng": 13.003822,
                  },
                  Object {
                    "lat": 56.381561,
                    "lng": 13.99286,
                  },
                ],
              },
              Object {
                "initialValue": Object {
                  "id": "cb228716-d3dd-444f-9a77-80443d436339",
                },
                "value": Object {
                  "id": "cb228716-d3dd-444f-9a77-80443d436339",
                },
              },
              Object {
                "initialValue": Array [
                  Object {
                    "id": "cb228716-d3dd-444f-9a77-80443d436339",
                  },
                  Object {
                    "id": "eb5732e2-b931-492b-82f1-f8fdd464f0d2",
                  },
                ],
                "value": Array [
                  Object {
                    "id": "cb228716-d3dd-444f-9a77-80443d436339",
                  },
                  Object {
                    "id": "eb5732e2-b931-492b-82f1-f8fdd464f0d2",
                  },
                ],
              },
              Object {
                "initialValue": Object {
                  "_type": "AnnotatedBar",
                  "annotation": "Annotation",
                  "bar": Object {
                    "id": "eb5732e2-b931-492b-82f1-f8fdd464f0d2",
                  },
                },
                "value": Object {
                  "_type": "AnnotatedBar",
                  "annotation": "Annotation",
                  "bar": Object {
                    "id": "eb5732e2-b931-492b-82f1-f8fdd464f0d2",
                  },
                },
              },
              Object {
                "initialValue": Array [
                  Object {
                    "_type": "AnnotatedBar",
                    "annotation": "First",
                    "bar": Object {
                      "id": "cb228716-d3dd-444f-9a77-80443d436339",
                    },
                  },
                  Object {
                    "_type": "AnnotatedBar",
                    "annotation": "Second",
                    "bar": Object {
                      "id": "eb5732e2-b931-492b-82f1-f8fdd464f0d2",
                    },
                  },
                ],
                "value": Array [
                  Object {
                    "_type": "AnnotatedBar",
                    "annotation": "First",
                    "bar": Object {
                      "id": "cb228716-d3dd-444f-9a77-80443d436339",
                    },
                  },
                  Object {
                    "_type": "AnnotatedBar",
                    "annotation": "Second",
                    "bar": Object {
                      "id": "eb5732e2-b931-492b-82f1-f8fdd464f0d2",
                    },
                  },
                ],
              },
              Object {
                "initialValue": null,
                "value": null,
              },
            ],
            "initialName": "Foo 1",
            "name": "Foo 1",
            "version": 0,
          },
          "entityLoadMessage": null,
          "id": "fc66b4d7-61ff-44d4-8f68-cb7f526df046",
          "initMessage": null,
        }
      `);
    }
  });
});
