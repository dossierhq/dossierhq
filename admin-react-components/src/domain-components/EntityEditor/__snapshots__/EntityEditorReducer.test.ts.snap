// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reduceEntityEditorState AddDraftAction new entity 1`] = `
Object {
  "activeEntityId": "e4e78fce-1089-41b8-9b6b-440d2e044061",
  "drafts": Array [
    Object {
      "entity": Object {
        "fields": Array [
          Object {
            "initialValue": null,
            "value": null,
          },
          Object {
            "initialValue": null,
            "value": null,
          },
          Object {
            "initialValue": null,
            "value": null,
          },
          Object {
            "initialValue": null,
            "value": null,
          },
          Object {
            "initialValue": null,
            "value": null,
          },
          Object {
            "initialValue": null,
            "value": null,
          },
          Object {
            "initialValue": null,
            "value": null,
          },
          Object {
            "initialValue": null,
            "value": null,
          },
          Object {
            "initialValue": null,
            "value": null,
          },
          Object {
            "initialValue": null,
            "value": null,
          },
        ],
        "initialName": "",
        "name": "",
        "version": 0,
      },
      "entityLoadMessage": null,
      "exists": false,
      "id": "e4e78fce-1089-41b8-9b6b-440d2e044061",
      "initMessage": null,
      "latestServerVersion": null,
      "publishState": null,
    },
  ],
}
`;

exports[`reduceEntityEditorState UpdateEntityAction 1`] = `
Object {
  "activeEntityId": "fc66b4d7-61ff-44d4-8f68-cb7f526df046",
  "drafts": Array [
    Object {
      "entity": Object {
        "fields": Array [
          Object {
            "initialValue": "Hello",
            "value": "Hello",
          },
          Object {
            "initialValue": Array [
              "one",
              "two",
              "three",
            ],
            "value": Array [
              "one",
              "two",
              "three",
            ],
          },
          Object {
            "initialValue": Object {
              "lat": 55.60498,
              "lng": 13.003822,
            },
            "value": Object {
              "lat": 55.60498,
              "lng": 13.003822,
            },
          },
          Object {
            "initialValue": Array [
              Object {
                "lat": 55.60498,
                "lng": 13.003822,
              },
              Object {
                "lat": 56.381561,
                "lng": 13.99286,
              },
            ],
            "value": Array [
              Object {
                "lat": 55.60498,
                "lng": 13.003822,
              },
              Object {
                "lat": 56.381561,
                "lng": 13.99286,
              },
            ],
          },
          Object {
            "initialValue": Object {
              "id": "cb228716-d3dd-444f-9a77-80443d436339",
            },
            "value": Object {
              "id": "cb228716-d3dd-444f-9a77-80443d436339",
            },
          },
          Object {
            "initialValue": Array [
              Object {
                "id": "cb228716-d3dd-444f-9a77-80443d436339",
              },
              Object {
                "id": "eb5732e2-b931-492b-82f1-f8fdd464f0d2",
              },
            ],
            "value": Array [
              Object {
                "id": "cb228716-d3dd-444f-9a77-80443d436339",
              },
              Object {
                "id": "eb5732e2-b931-492b-82f1-f8fdd464f0d2",
              },
            ],
          },
          Object {
            "initialValue": Object {
              "blocks": Array [
                Object {
                  "data": Object {
                    "text": "Hello world",
                  },
                  "type": "paragraph",
                },
              ],
            },
            "value": Object {
              "blocks": Array [
                Object {
                  "data": Object {
                    "text": "Hello world",
                  },
                  "type": "paragraph",
                },
              ],
            },
          },
          Object {
            "initialValue": Object {
              "_type": "AnnotatedBar",
              "annotation": "Annotation",
              "bar": Object {
                "id": "eb5732e2-b931-492b-82f1-f8fdd464f0d2",
              },
            },
            "value": Object {
              "_type": "AnnotatedBar",
              "annotation": "Annotation",
              "bar": Object {
                "id": "eb5732e2-b931-492b-82f1-f8fdd464f0d2",
              },
            },
          },
          Object {
            "initialValue": Array [
              Object {
                "_type": "AnnotatedBar",
                "annotation": "First",
                "bar": Object {
                  "id": "cb228716-d3dd-444f-9a77-80443d436339",
                },
              },
              Object {
                "_type": "AnnotatedBar",
                "annotation": "Second",
                "bar": Object {
                  "id": "eb5732e2-b931-492b-82f1-f8fdd464f0d2",
                },
              },
            ],
            "value": Array [
              Object {
                "_type": "AnnotatedBar",
                "annotation": "First",
                "bar": Object {
                  "id": "cb228716-d3dd-444f-9a77-80443d436339",
                },
              },
              Object {
                "_type": "AnnotatedBar",
                "annotation": "Second",
                "bar": Object {
                  "id": "eb5732e2-b931-492b-82f1-f8fdd464f0d2",
                },
              },
            ],
          },
          Object {
            "initialValue": null,
            "value": null,
          },
        ],
        "initialName": "Foo 1",
        "name": "Foo 1",
        "version": 1,
      },
      "entityLoadMessage": null,
      "exists": true,
      "id": "fc66b4d7-61ff-44d4-8f68-cb7f526df046",
      "initMessage": null,
      "latestServerVersion": 0,
      "publishState": "published",
    },
  ],
}
`;
